packages/core/src/Room.ts

```diff
@@ -599,18 +614,23 @@ export abstract class Room<State extends object= any, Metadata= any> {
       }
     }
 
-    // emit 'join' to room handler
-    this._events.emit('join', client);
+    // emit 'join' to room handler (if not reconnecting)
+    if (!previousReconnectionToken) {
+      this._events.emit('join', client);
+    }
 
-    // allow client to send messages after onJoin has succeeded.
-    client.ref.on('message', this._onMessage.bind(this, client));
+    // state might already be ClientState.LEAVING here
+    if (client.state === ClientState.JOINING) {
```

Comment:
does this work during RECONNECT? It looked to me like the client.state would be RECONNECTING, and the code below needs to run when reconnecting as well as when JOINING

Reply:
if state before calling is changed to JOINING, however, it should still work.

Reply:
Yes! In fact the `ClientState.RECONNECTED` is only used to prevent the `'leave'` event from being emitted on the first connection after the reconnection has been confirmed. https://github.com/colyseus/colyseus/blob/4ff32c4fe2b143538ac75a6823455c5835f135a5/packages/core/src/Room.ts#L940-L945

Reply:
Sorry, I meant when reconnecting, the state may be LEAVING or JOINED, because it never gets set back to JOINING after it gets set to JOINED.
I think I explicitly set it to JOINING in beginning of that function, but I didn't see if you copied that.